### PR TITLE
dev-57: Multithreading support

### PR DESCRIPTION
**Pull request by StijnKlopper**
Added multithreading support
Multithreading explanation tilebuilder/cavebuilder:
1.  Initiate starts a thread with a callback function.
2.  Inside the thread tile data is being generated.
3.  Once the thread is finished we add the data along with the callback to a concurrent queue.
4.  Inside the update function (Main Thread) of the tile/cavebuilder we check every frame to see if something has been added to the queue.
5.  If the queue is not empty we dequeue and obtain the callback and data.
6.  We call the callback function and pass the data as a paramter.
7.  The callback function is a OnDataReceive function that handles setting all the Unity Specific data inside the main thread.

Various changes:
- Tile builder and cave builder are no longer part of seperate gameobjects but only created once
- Objectpool no longer creates seperate terrain, cave and ocean gameobject but creates tile object which has these gameobjects included.
- Tile class has been heavily altered.
- cavebuilder is no longer being called from worldbuilder but from tilebuilder similair to city generator. This is to make sure tilebuilder has generated the heightmap that cavebuilder needs.
- Cleaned up cavebuilder, removed excess code and improved garbage collection problem which was creating too many new vector3 arrays which caused significant slowdown.

### CODE REVIEW

**Review by Tarikyil1998 — APPROVED** [review]

File: No FYS-X Sky/Assets/Helpers/SafeMesh.cs [hunk 1]
```diff
@@ -4,6 +4,8 @@ public class SafeMesh
 {
     public int[] Triangles;
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/Helpers/SafeMesh.cs:6]: Remove whitespace

File: No FYS-X Sky/Assets/Helpers/SafeMesh.cs [hunk 2]
```diff
@@ -4,6 +4,8 @@ public class SafeMesh
...
 
+    public Vector3 position;
+
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/Helpers/SafeMesh.cs:8]: Remove whitespace

File: No FYS-X Sky/Assets/Helpers/SafeMesh.cs [hunk 3]
```diff
@@ -4,6 +4,8 @@ public class SafeMesh
...
     public Vector3[] Vertices;
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/Helpers/SafeMesh.cs:10]: Remove whitespace

File: No FYS-X Sky/Assets/World/Builders/CaveBuilder.cs [hunk 4]
```diff
@@ -1,76 +1,82 @@
...
 using UnityEngine;
 
 public class CaveBuilder : MonoBehaviour
 {
-
     private RidgedMultifractal ridgedMultifractal;
-    private SafeMesh safeMesh;
-    private float[,] heightMap;
+    private ConcurrentQueue<CaveThreadInfo<SafeMesh>> caveDataThreadInfoQueue;
+    public const int CAVE_DEPTH = 30;
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/World/Builders/CaveBuilder.cs:11]: Const should be on top of the list of vars

File: No FYS-X Sky/Assets/World/Builders/CaveBuilder.cs [hunk 5]
```diff
@@ -1,76 +1,82 @@
...
+        caveDataThreadInfoQueue.Enqueue(new CaveThreadInfo<SafeMesh>(callback, safeMesh));
     }
 
-    private IEnumerator GenerateCaveMap(Mesh caveMesh, Vector2 offsets, int height)
+    private void OnCaveDataReceived(SafeMesh safeMesh)
     {
-        safeMesh = this.GenerateCaveMap(offsets, caveMesh, height);
-        yield return safeMesh;
+        Tile currentTile = WorldBuilder.GetTile(safeMesh.position);
+        // check if currentTile is not already unloaded
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/World/Builders/CaveBuilder.cs:43]: The comment should start with a capital and there should be a white line before the comment.

File: No FYS-X Sky/Assets/World/Builders/CaveBuilder.cs [hunk 6]
```diff
@@ -1,76 +1,82 @@
...
+        if (currentTile != null) 
+        {
+            Mesh mesh = new Mesh();
+            mesh.vertices = safeMesh.Vertices;
+            mesh.triangles = safeMesh.Triangles;
+            currentTile.Cave.MeshFilter.mesh = mesh;
+            mesh.uv = GenerateUV.CalculateUVs(safeMesh.Vertices, 1);
+            mesh.RecalculateNormals();
+
+            //Find out what this does
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/World/Builders/CaveBuilder.cs:53]: Please place a space between // and Find, like:
// Find out what this does

File: No FYS-X Sky/Assets/World/Builders/TileBuilder.cs [hunk 7]
```diff
@@ -1,101 +1,55 @@
...
-
     private CityGenerator cityGenerator;
-
-
-    private float[] tileTextureData;
-
     private Texture2DArray splatmapsArray;
+    private CaveBuilder caveBuilder;
+    private ConcurrentQueue<MapThreadInfo<TileData>> terrainDataThreadInfoQueue;
+    public const int TILE_DIMENSION = 11;
```
Comment (appelmoesje) [inline No FYS-X Sky/Assets/World/Builders/TileBuilder.cs:14]: Const should be on top, like the other review.

**Review by appelmoesje — CHANGES_REQUESTED** [review]

**Review by appelmoesje — APPROVED** [review]

**Pull request merged**